tools/c7n_policystream/README.md

```diff
@@ -0,0 +1,93 @@
+# Policy Changes from Git
+
+Using custodian in accordance with infrastructure as code principles,
+we store policy assets in a versioned control repository. This
+provides for an audit log and facilitates code reviews. However this
+capability is primarily of use to humans making semantic interpretations
+of changes.
+
+This script also provides logical custodian policy changes over a git
+repo and allows streaming those changes for machine readable/application
+consumption. Its typically used as a basis for CI integrations or indexes
+over policies.
+
+Two example use cases:
+
+  - Doing dryrun only on changed policies within a pull request
+  - Constructing a database of policy changes.
+
+Policystream works on individual github repositories, or per Github integration
+across an organization's set of repositories.
+
+# Install
+
+policystream can be installed via pypi, provided the require pre-requisites
+libraries are available (libgit2 > 0.26)
+
+```
+pip install c7n
+```
+
+Docker images available soon, see build for constructing your own.
+
+# Build
+
+Alternatively a docker image can be built as follows
+
+```shell
+# Note must be top level directory of checkout
+cd cloud-custodian
+docker build -t policystream:latest -f tools/c7n_policystream/Dockerfile .
+```
+
+# Usage
+
+Streaming use case (default stream is to stdout, also supports kinesis, rdbms and sqs)
+
+```
+  $ python tools/ops/policystream.py stream -r foo
+  2018-08-12 12:37:00,567: c7n.policystream:INFO Cloning repository: foo
+  <policy-add policy:foi provider:aws resource:ec2 date:2018-08-02T15:13:28-07:00 author:Kapil commit:09cb85>
+  <policy-moved policy:foi provider:aws resource:ec2 date:2018-08-02T15:14:24-07:00 author:Kapil commit:76fce7>
+  <policy-remove policy:foi provider:aws resource:ec2 date:2018-08-02T15:14:46-07:00 author:Kapil commit:570ca4>
+  <policy-add policy:ec2-guard-duty provider:aws resource:ec2 date:2018-08-02T15:14:46-07:00 author:Kapil commit:570ca4>
+  <policy-add policy:ec2-run provider:aws resource:ec2 date:2018-08-02T15:16:00-07:00 author:Kapil commit:d3d8d4>
+  <policy-remove policy:ec2-run provider:aws resource:ec2 date:2018-08-02T15:18:31-07:00 author:Kapil commit:922c1a>
+  <policy-modified policy:ec2-guard-duty provider:aws resource:ec2 date:2018-08-12T09:39:43-04:00 author:Kapil commit:189ea1>
+  2018-08-12 12:37:01,275: c7n.policystream:INFO Streamed 7 policy changes
+```
+
+Diff use case, output policies changes in the last commit
+
+```
+  $ python tools/ops/policystream.py diff -r foo -v
```

Comment:
I was not able to get a diff from the last commit using this example. If I specify the commit SHA1 as the `--source` then it shows the diff.

```python tools/ops/policystream.py diff -r . --source <SHA1> --target master```
works

Reply:
yeah, i should clarify that its really diff'ing current head symbolic to master symbolic pointer, the intent is that your on a branch and diffing it to master like in a pr check, i could probably add some logic to switch it to HEAD^ to HEAD if HEAD == MASTER.

Reply:
The scenario I'm trying is once the PR is merged to master I want to have CI find only the diffs from the merge/commit. CI would then push only the changed policies to the release pipeline.

Reply:
The scenario I'm trying is once the PR is merged to master I want to have CI find only the diffs from the merge/commit. CI would then push only the changed policies to the release pipeline.

Reply:
so source/target are cli parameters to diff, you can explicitly pass source and target references (and it supports various fancy rev parsing). See https://git-scm.com/book/en/v2/Git-Tools-Revision-Selection

The issue with doing something `smart` for that  use case is that afaics its dependent on workflow, ie. squash and merge would just be head to head^  but a normal merge means walking back to the last commit with multiple parents. even detecting master is tricky as the name is just a convention. It does look like git/libgit can support that distinction for us via {1} syntax.

i've gone ahead with the logic to support both the use cases (on master show diff to last and on branch show diff to master), with tests, and additional cli and readme docs on the defaults.